### PR TITLE
Add a fakeroot engine for build

### LINKED ISSUE #3892: building definition file directly to sandbox with --fakeroot fail.
## Version of Singularity:

3.3.0-rc.1.142

### Expected behavior

Building sandbox directly from definition file failed,
but if building from the image .sif file created from the def file it succeed.

**$ singularity build --fakeroot simg.sif Singularity.def**
INFO:    Starting build...
...
INFO:    Creating SIF file...
INFO:    Build complete: simg.sif

**$ singularity build --fakeroot --sandbox sandbox_chroot/ simg.sif** 
INFO:    Starting build...
INFO:    Creating sandbox directory...
INFO:    Build complete: sandbox_chroot/

**$ singularity build --fakeroot --sandbox sandbox_chroot/ Singularity.def**
...
FATAL:   While performing build: Sandbox Assemble Failed: exit status 1: mv: cannot access '/tmp/sbuild-928059436/fs/var/cache/apt/archives/partial': Permission denied

would like to build to the sandbox directly from def file and succeed with --fakeroot.

### Actual behavior

$ **singularity -d build --fakeroot --sandbox sandbox_chroot/ Singularity.def** 
DEBUG   [U=12544,P=16855]createConfDir()               /nfs/homes/user/.singularity already exits. Not creating.
DEBUG   [U=12544,P=16855]getCacheBasedir()             environment variable SINGULARITY_CACHEDIR not set, using default image cache
...
DEBUG   [U=12544,P=16855]bundleCommon()                Created temporary directory for bundle /tmp/sbuild-782313980
INFO    [U=12544,P=16855]Full()                        Starting build...
...
DEBUG   [U=12544,P=16934]Master()                      Child exited with exit status 0
DEBUG   [U=12544,P=16855]Full()                        Inserting Metadata
DEBUG   [U=12544,P=16855]Full()                        Calling assembler
INFO    [U=12544,P=16855]Assemble()                    Creating sandbox directory...
DEBUG   [U=12544,P=16855]Assemble()                    Moving sandbox from /tmp/sbuild-782313980/fs to sandbox_chroot/
DEBUG   [U=12544,P=16855]cleanUp()                     Build bundle(s) cleaned: [/tmp/sbuild-782313980]
**FATAL   [U=12544,P=16855]run()                         While performing build: Sandbox Assemble Failed: exit status 1: mv: cannot access '/tmp/sbuild-782313980/fs/var/cache/apt/archives/partial': Permission denied
mv: cannot access '/tmp/sbuild-782313980/fs/var/lib/apt/lists/partial': Permission denied**

### Steps to reproduce behavior

Singularity.def:
\------------------
BootStrap: docker
From: ubuntu:18.04

%post
  apt-get update
  apt-get -y install lsb-release
\------------------
$ **singularity -d build --fakeroot --sandbox sandbox_chroot/ Singularity.def**

### LINKED ISSUE #3927: Check for ability to use --fakeroot earlier
## Version of Singularity:

3.3.0-rc.1.204.gb5e5078a9

### Expected behavior
```
$ singularity build --fakeroot test.sif test.def 
FATAL:   you are not allowed to use fakeroot
```
### Actual behavior

```
$ singularity build --fakeroot test.sif test.def 
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:5b7339215d1d5f8e68622d584a224f60339f5bef41dbd74330d081e912f0cddd
 25.45 MiB / 25.45 MiB [====================================================] 2s
Copying blob sha256:14ca88e9f6723ce82bc14b241cda8634f6d19677184691d086662641ab96fe68
 34.52 KiB / 34.52 KiB [====================================================] 0s
Copying blob sha256:a31c3b1caad473a474d574283741f880e37c708cc06ee620d3e93fa602125ee0
 849 B / 849 B [============================================================] 0s
Copying blob sha256:b054a26005b7f3b032577f811421fab5ec3b42ce45a4012dfa00cf6ed6191b0f
 162 B / 162 B [============================================================] 0s
Copying config sha256:84c9d0762469176a58d9c54375a1e9a4dcdc0045e3e14e191d97bd12cd5b23a1
 2.42 KiB / 2.42 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
FATAL:   you are not allowed to use fakeroot
FATAL:   While performing build: while running engine: exit status 255
```
We could have avoided downloading all of the blobs if we had just checked to see if the user was permitted to use `--fakeroot` in the CLI before initiating anything else.  Note that is not great, because the blobs end up in the cache.  
```
$ singularity cache list --verbose --type blobs
NAME                     DATE CREATED           SIZE             TYPE
0b63974fc3d2b5516cc72c   2019-07-12 10:30:28    0.81 kB          blob
14ca88e9f6723ce82bc14b   2019-07-12 10:30:26    35.35 kB         blob
5b7339215d1d5f8e68622d   2019-07-12 10:30:26    26.68 MB         blob
84c9d0762469176a58d9c5   2019-07-12 10:30:28    2.48 kB          blob
a31c3b1caad473a474d574   2019-07-12 10:30:27    0.85 kB          blob
b054a26005b7f3b032577f   2019-07-12 10:30:27    0.16 kB          blob

There are 6 oci blob file(s) using 26.72 MB of space
Total space used: 26.72 MB
```

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

- Create a fakeroot engine to fix build issues with fakeroot option
- Fix yum bootstrap with fakeroot.

**This fixes or addresses the following GitHub issues:**

- Fixes #3892
- Fixes #3927

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/pkg/build/sources/conveyorPacker_yum.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 	"regexp"
 	"runtime"
 	"strings"
+	"syscall"
```
Comment (WestleyK) [inline internal/pkg/build/sources/conveyorPacker_yum.go:20]: Copyright in this file should be `2018-2019`

File: internal/pkg/build/sources/conveyorPacker_yum.go [hunk 2]
```diff
@@ -305,25 +306,31 @@ func (c *YumConveyor) importGPGKey() (err error) {
...
-			if err != nil {
-				return fmt.Errorf("While creating %v: %v", filepath.Join(c.b.Rootfs(), dev), err)
-			}
-
-			defer f.Close()
+		d := int((dev.major << 8) | (dev.minor & 0xff) | ((dev.minor & 0xfff00) << 12))
+		path := filepath.Join(c.b.Rootfs(), dev.path)
+
+		if err := syscall.Mknod(path, dev.mode, d); err != nil {
+			return fmt.Errorf("While creating %s: %s", path, err)
```
Comment (WestleyK) [inline internal/pkg/build/sources/conveyorPacker_yum.go]:
```suggestion
			return fmt.Errorf("while creating %s: %s", path, err)
```

Not necessary, theres lot of other error messages that use caps. We can fix them all later.

File: internal/pkg/runtime/engines/fakeroot/engine_linux.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/fakeroot/engine_linux.go]: Extra newline?

File: internal/pkg/runtime/engines/fakeroot/engine_linux.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/opencontainers/runtime-tools/generate"
+
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/fakeroot/engine_linux.go]: Here too?

File: internal/pkg/runtime/engines/fakeroot/engine_linux.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	fakerootUtil "github.com/sylabs/singularity/internal/pkg/fakeroot"
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines/config"
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines/config/starter"
+	fakeroot "github.com/sylabs/singularity/internal/pkg/runtime/engines/fakeroot/config"
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/fakeroot/engine_linux.go]: Should this be something like `fakerootconf`?

```suggestion
	fakerootconf "github.com/sylabs/singularity/internal/pkg/runtime/engines/fakeroot/config"
```

**Review by WestleyK — COMMENTED** [review]
Couple minor suggestions. Otherwise looks good :+1:

**Review by ikaneshiro — APPROVED** [review]
lgtm

File: internal/pkg/runtime/engines/fakeroot/engine_linux.go (hunk 5 above)
Comment (cclerget) [reply internal/pkg/runtime/engines/fakeroot/engine_linux.go]: fixed

**Pull request merged**